### PR TITLE
fix: Negative Qty and Rates in SO/PO

### PR DESCRIPTION
**Pull request by marination**
**Issue:**
- This is allowed in Sales and Purchase Orders
   Screenshot 2023-11-21 at 1 29 19 PM

- The same isn't allowed via Update Items
  Screenshot 2023-11-21 at 1 30 05 PM
- Similarly negative rates aren't allowed via Update Items either

**Fix:**
- Make sure on SO/PO save the **qty** is validated (negative qty not allowed)
	Screenshot 2023-11-21 at 1 27 13 PM
- Also, allow negative rates on Update Items and normal doc submission/save

### CODE REVIEW

File: erpnext/controllers/accounts_controller.py [hunk 1]
```diff
@@ -71,6 +71,10 @@ class AccountMissingError(frappe.ValidationError):
 	pass
 
 
+class InvalidQtyError(frappe.ValidationError):
+	pass
```
Comment (bosue) [inline erpnext/controllers/accounts_controller.py:75]: Nice, I’ll use that one in #37865, too. That‘ll simplify things a bit.

File: erpnext/controllers/accounts_controller.py [hunk 2]
```diff
@@ -910,10 +914,23 @@ def get_value_in_transaction_currency(self, account_currency, args, field):
...
+		for item in self.items:
+			if not flt(item.qty):
+				frappe.throw(
+					msg=_("Row #{0}: Item quantity can not be zero").format(item.idx),
+					title=title,
+					exc=InvalidQtyError,
+				)
+			if flt(item.qty) < 0.0 and self.doctype in ["Purchase Order", "Sales Order"]:
+				frappe.throw(
+					msg=_("Row #{0}: Item quantity can not be negative").format(item.idx),
```
Comment (bosue) [inline erpnext/controllers/accounts_controller.py]: English grammar is a bit weird, so “cannot” is written as one word… 😅

Comment (marination) [inline erpnext/controllers/accounts_controller.py]:
```suggestion
					msg=_("Row #{0}: Item quantity cannot be zero").format(item.idx),
					title=title,
					exc=InvalidQtyError,
				)
			if flt(item.qty) < 0.0 and self.doctype in ["Purchase Order", "Sales Order"]:
				frappe.throw(
					msg=_("Row #{0}: Item quantity cannot be negative").format(item.idx),
```

File: erpnext/controllers/accounts_controller.py [hunk 3]
```diff
@@ -910,10 +914,23 @@ def get_value_in_transaction_currency(self, account_currency, args, field):
...
+					msg=_("Row #{0}: Item quantity cannot be zero").format(item.idx),
+					title=title,
+					exc=InvalidQtyError,
+				)
+			if flt(item.qty) < 0.0 and self.doctype in ["Purchase Order", "Sales Order"]:
+				frappe.throw(
+					msg=_("Row #{0}: Item quantity cannot be negative").format(item.idx),
+					title=title,
+					exc=InvalidQtyError,
+				)
```
Comment (rohitwaghchaure) [inline erpnext/controllers/accounts_controller.py:926]: @marination can we use this property for SO and PO? 

Screenshot 2023-11-23 at 10 45 03 PM

Comment (marination) [reply erpnext/controllers/accounts_controller.py:926]: done

**Pull request merged**